File: rust/e2e/tests/test_burst_gpt_trace.rs
// SPDX-FileCopyrightText: Copyright (c) 2026 NVIDIA CORPORATION & AFFILIATES. All rights reserved.
// SPDX-License-Identifier: Apache-2.0
mod common;
use common::*;

use std::path::{Path, PathBuf};

fn sample_rows() -> Vec<Vec<String>> {
    let raw: &[(&str, &str, u32, u32, u32, &str)] = &[
        ("0.0", "ChatGPT", 472, 18, 490, "Conversation log"),
        ("0.1", "ChatGPT", 1087, 230, 1317, "Conversation log"),
        ("0.2", "GPT-4", 417, 276, 693, "Conversation log"),
        ("0.3", "ChatGPT", 1360, 647, 2007, "Conversation log"),
        ("0.4", "ChatGPT", 185, 215, 400, "Conversation log"),
    ];
    raw.iter()
        .map(|(ts, model, req, resp, total, log)| {
            vec![
                ts.to_string(),
                model.to_string(),
                req.to_string(),
                resp.to_string(),
                total.to_string(),
                log.to_string(),
            ]
        })
        .collect()
}

fn create_burst_gpt_csv_file(dir: &Path, rows: &[Vec<String>]) -> PathBuf {
    let headers = [
        "Timestamp",
        "Model",
        "Request tokens",
        "Response tokens",
        "Total tokens",
        "Log Type",
    ];
    write_csv(dir, "burst_gpt.csv", &headers, rows)
}

#[tokio::test]
async fn test_fixed_schedule_with_explicit_dataset_type() {
    let h = AIPerfHarness::new().await;
    let rows = sample_rows();
    let csv_file = create_burst_gpt_csv_file(h.artifact_dir.path(), &rows);

    let r = h.run(&format!(
        "--model {DEFAULT_MODEL} --url {} --endpoint-type chat \
         --input-file {} --custom-dataset-type burst_gpt_trace \
         --request-count {} --fixed-schedule --fixed-schedule-auto-offset \
         --workers-max 1 --ui simple",
        h.mock.url,
        csv_file.display(),
        rows.len(),
    ));

    assert_eq!(r.exit_code, 0);
    assert_eq!(r.artifacts.request_count() as usize, rows.len());
}

#[tokio::test]
async fn test_fixed_schedule_auto_detected() {
    let h = AIPerfHarness::new().await;
    let rows = sample_rows();
    let csv_file = create_burst_gpt_csv_file(h.artifact_dir.path(), &rows);

    let r = h.run(&format!(
        "--model {DEFAULT_MODEL} --url {} --endpoint-type chat \
         --input-file {} --request-count {} \
         --fixed-schedule --fixed-schedule-auto-offset \
         --workers-max 1 --ui simple",
        h.mock.url,
        csv_file.display(),
        rows.len(),
    ));

    assert_eq!(r.exit_code, 0);
    assert_eq!(r.artifacts.request_count() as usize, rows.len());
}

File: tests/integration/test_burst_gpt_trace.py
# SPDX-FileCopyrightText: Copyright (c) 2026 NVIDIA CORPORATION & AFFILIATES. All rights reserved.
# SPDX-License-Identifier: Apache-2.0

"""Integration tests for BurstGPT trace custom dataset type."""

from pathlib import Path

import pytest

from tests.harness.utils import AIPerfCLI, AIPerfMockServer
from tests.integration.conftest import IntegrationTestDefaults as defaults
from tests.integration.utils import create_burst_gpt_csv_file


def _sample_rows() -> list[dict]:
    """A trimmed BurstGPT-shaped CSV with sub-second timestamps."""
    return [
        {"Timestamp": 0.0, "Model": "ChatGPT", "Request tokens": 472, "Response tokens": 18, "Total tokens": 490, "Log Type": "Conversation log"},
        {"Timestamp": 0.1, "Model": "ChatGPT", "Request tokens": 1087, "Response tokens": 230, "Total tokens": 1317, "Log Type": "Conversation log"},
        {"Timestamp": 0.2, "Model": "GPT-4", "Request tokens": 417, "Response tokens": 276, "Total tokens": 693, "Log Type": "Conversation log"},
        {"Timestamp": 0.3, "Model": "ChatGPT", "Request tokens": 1360, "Response tokens": 647, "Total tokens": 2007, "Log Type": "Conversation log"},
        {"Timestamp": 0.4, "Model": "ChatGPT", "Request tokens": 185, "Response tokens": 215, "Total tokens": 400, "Log Type": "Conversation log"},
    ]  # fmt: skip


@pytest.mark.integration
@pytest.mark.asyncio
class TestBurstGPTTraceIntegration:
    """Integration tests for burst_gpt_trace dataset loader."""

    async def test_fixed_schedule_with_explicit_dataset_type(
        self,
        cli: AIPerfCLI,
        aiperf_mock_server: AIPerfMockServer,
        tmp_path: Path,
    ) -> None:
        """``--custom-dataset-type burst_gpt_trace --fixed-schedule`` runs end-to-end."""
        rows = _sample_rows()
        csv_file = create_burst_gpt_csv_file(tmp_path, rows)

        result = await cli.run(
            f"""
            aiperf profile \
                --model {defaults.model} \
                --url {aiperf_mock_server.url} \
                --endpoint-type chat \
                --input-file {csv_file} \
                --custom-dataset-type burst_gpt_trace \
                --request-count {len(rows)} \
                --fixed-schedule \
                --fixed-schedule-auto-offset \
                --workers-max {defaults.workers_max} \
                --ui {defaults.ui}
            """
        )

        assert result.exit_code == 0
        assert result.request_count == len(rows)
        assert result.has_all_outputs_except_inputs
        assert result.inputs is None, "trace datasets skip inputs.json"

    async def test_fixed_schedule_auto_detected(
        self,
        cli: AIPerfCLI,
        aiperf_mock_server: AIPerfMockServer,
        tmp_path: Path,
    ) -> None:
        """No ``--custom-dataset-type`` flag — the loader's ``can_load``"""
        rows = _sample_rows()
        csv_file = create_burst_gpt_csv_file(tmp_path, rows)

        result = await cli.run(
            f"""
            aiperf profile \
                --model {defaults.model} \
                --url {aiperf_mock_server.url} \
                --endpoint-type chat \
                --input-file {csv_file} \
                --request-count {len(rows)} \
                --fixed-schedule \
                --fixed-schedule-auto-offset \
                --workers-max {defaults.workers_max} \
                --ui {defaults.ui}
            """
        )

        assert result.exit_code == 0
        assert result.request_count == len(rows)
        assert result.has_all_outputs_except_inputs
        assert result.inputs is None, "trace datasets skip inputs.json"
